SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift

```diff
@@ -48,28 +50,32 @@ private func optimizeFunctionsTopDown(using worklist: inout FunctionWorklist,
 }
 
 private func optimize(function: Function, _ context: FunctionPassContext) {
-  runSimplification(on: function, context, preserveDebugInfo: true) { instruction, simplifyCtxt in
-    if let i = instruction as? OnoneSimplifyable {
-      i.simplify(simplifyCtxt)
-      if instruction.isDeleted {
-        return
+  var changed = true
+  while changed {
+    changed = false
+    
+    changed = changed || runSimplification(on: function, context, preserveDebugInfo: true) { instruction, simplifyCtxt in
```

Comment:
Is it really required to retrigger another iteration for all the things we do here?
I think it's only required if `optimizeMemoryAccesses` does some changes. Only this optimization can open more simplification opportunities.

Reply:
It seems like I need to at least check if runSimplification did anything *or* optimizeMemoryAccesses did anything, it's not enough to just check optimizeMemoryAccesses. Does that sound about right? I can find out why exactly.